source/common/tcp_proxy/upstream.h

```diff
@@ -68,16 +86,22 @@ class HttpConnPool : public GenericConnPool, public Http::ConnectionPool::Callba
                    Upstream::HostDescriptionConstSharedPtr host, StreamInfo::StreamInfo& info,
                    absl::optional<Http::Protocol>) override;
 
+  void onUpstreamHostSelected(Upstream::HostDescriptionConstSharedPtr);
+  void onHttpPoolReady(Upstream::HostDescriptionConstSharedPtr& host,
+                       Ssl::ConnectionInfoConstSharedPtr ssl_info);
+
   class Callbacks {
   public:
     Callbacks(HttpConnPool& conn_pool, Upstream::HostDescriptionConstSharedPtr host,
               Ssl::ConnectionInfoConstSharedPtr ssl_info)
         : conn_pool_(&conn_pool), host_(host), ssl_info_(ssl_info) {}
     virtual ~Callbacks() = default;
-    virtual void onSuccess(Http::RequestEncoder& request_encoder) {
+    virtual void onSuccess(Http::RequestEncoder* request_encoder) {
       ASSERT(conn_pool_ != nullptr);
-      conn_pool_->onGenericPoolReady(host_, request_encoder.getStream().connectionInfoProvider(),
-                                     ssl_info_);
+      if (request_encoder != nullptr) {
```

Comment:
request_encoder will be null in case CombinedUpstream is handling the stream.  tcp_proxy will not be managing request_encoder in case of CombinedStream/Router::UpstreamRequest.  In this PR, rest of the changes I skipped because that would have required feature gate. Please refer[ here](https://github.com/envoyproxy/envoy/pull/27183/files#diff-8b2f11a21a008c86b7a6203cd94bd8b94455fe9ca4e9c04a52537c0233b3b9afR316) in the parent PR for complete changes